TaboolaSDKExampleSwift/TaboolaSDKExampleSwift/SDK Native/NativeViewController.swift

```diff
@@ -10,60 +10,110 @@ import Foundation
 }
 
-extension NativeViewController: UITableViewDataSource {
-    func numberOfSections(in tableView: UITableView) -> Int {
+extension NativeViewController: UICollectionViewDelegate, UICollectionViewDataSource {
+    func numberOfSections(in collectionView: UICollectionView) -> Int {
         return Constants.totalSectionsNative
     }
     
-    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
-        return Constants.totalRowsNative
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        guard let itemsArray = itemsArray else { return 0 }
+        return itemsArray.count
     }
     
-    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
-        let cell = tableView.dequeueReusableCell(withIdentifier: "TextCell") ?? UITableViewCell(style: .default, reuseIdentifier: "TextCell")
-        cell.textLabel?.text = Constants.nativeText
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "NativeCollectionViewCell", for: indexPath)
+        if let cell = cell as? NativeViewControllerCell {
+            cell.attributionButton.addTarget(self, action: #selector(clickedTaboolaAttribution), for: .touchDown)
+            guard let itemsArray = itemsArray else { return cell }
+            if let item = itemsArray[indexPath.row] as? TBLNativeItem {
+                if let imageUrl = item.extraDataDictionary()["imageUrl"] as? NSString {
+                    DispatchQueue.global(qos: .background).async {
+                        guard let data = NSData.init(contentsOf: NSURL(string: imageUrl as String)! as URL) else { return }
+                        DispatchQueue.main.async {
+                            cell.imageView.image = UIImage(data: data as Data)
+                        }
+                    }
+                    cell.imageView.isHidden = false
+                    cell.tbImageView.isHidden = true
+                    cell.tbImageView.isUserInteractionEnabled = false
+                    cell.imageView.isUserInteractionEnabled = true
+                } else {
+                    item.initThumbnailView(cell.tbImageView)
+                    cell.imageView.isHidden = true
+                    cell.tbImageView.isHidden = false
+                    cell.tbImageView.isUserInteractionEnabled = true
+                    cell.imageView.isUserInteractionEnabled = false
+                }
+                
+                item.initTitleView(cell.titleLabel)
+                item.initBrandingView(cell.brandingLabel)
+            }
+        }
         return cell
     }
+    
+    func collectionView(_ collectionView: UICollectionView, willDisplay cell: UICollectionViewCell, forItemAt indexPath: IndexPath) {
+        guard let itemsArray = itemsArray else { return }
```

Comment:
In here and in other places, you're usually dealing only with the happy scenario. What happens if the flow fails? At least write a log